earth_enterprise/src/common/khMTTypes.h

```diff
@@ -148,9 +151,173 @@ class khMTSet
+      thekeys.push_back(p.first);
+    }
+    return thekeys;
+  }
+  bool contains(const K& key) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map.find(key) != map.end();
+  }
+  bool empty(void) const {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map.empty();
+  }
+  khMTMap(const khMTMap&) {
+    // todo: this is only used for setting up an empty dirtymap by AssetD and AssetVersionD
+    // should probably really remove this
+  }
+};
+
+// WARNING: Use this with extreme caution
+// It protects some obvious problems but inheriting from vector publicly isn't great
+template <typename T>
+class MTVector : public std::vector<T> {
+  private:
+    mutable std::mutex mtx;
+
+    MTVector(const MTVector& a, const std::lock_guard<std::mutex> &) : Base(a) { }
+  public:
+    typedef std::vector<T> Base;
+    typedef typename Base::const_iterator const_iterator;
+    void push_back(const T& v) {
+      std::lock_guard<std::mutex> lock(mtx);
+      Base::push_back(v);
+    }
+    MTVector() : Base() { }
+    // forward to private copy constructor to protect vector from modification
+    MTVector(const MTVector& a) : MTVector(a, std::lock_guard<std::mutex>(a.mtx)) { }
+    const T& operator[] (const size_t idx) const {
+      std::lock_guard<std::mutex> lock(mtx);
+      return Base::operator[](idx); 
+    }
+
+    bool operator==(const MTVector& x) const {
+      std::lock_guard<std::mutex> lock(mtx);
+      return true;
+    }
+
+    bool operator==(const Base& x) const {
+      std::lock_guard<std::mutex> lock(mtx);
+      return true;
+    }
+
+    void clear(void) {
+      std::lock_guard<std::mutex> lock(mtx);
+      Base::clear();
+    }
+
+    void shrink_to_fit(void) {
+      std::lock_guard<std::mutex> lock(mtx);
+      Base::shrink_to_fit();
+    }
+
+    const_iterator begin(void) const {
+      return Base::begin();
+    }
+
+    const_iterator end(void) const {
+      return Base::end();
+    }
+
+    size_t size(void) const {
+      std::lock_guard<std::mutex> lock(mtx);
+      return Base::size();
+    }
+};
+
+class khTargetedLock {
+  private:
+    // no such thing as a targeted lock which doesn't have a target
+    khTargetedLock() = delete;
+    khTargetedLock(const khTargetedLock&) = delete;
+    typedef std::unordered_map<std::string, std::shared_ptr<std::mutex>> FileLocks;
+    
+    static std::mutex& globalMutex(void) {
+      static std::mutex classMutex;
+      return classMutex;
+    }
+    std::shared_ptr<std::mutex> theFileMutex;
+    const std::string& s;
+    static FileLocks& getFileLocks(void) {
+      static FileLocks file_locks;
+      return file_locks;
+    } 
+  public:
+    khTargetedLock(const std::string& filename) : s(filename) {
+      FileLocks& file_mutexes = getFileLocks();
+      {
+        std::lock_guard<std::mutex> global_lock(globalMutex());
+        if (file_mutexes.find(s) == file_mutexes.end()) {
+          file_mutexes[s] = std::make_shared<std::mutex>();
```

Comment:
I don't think there's a need for a shared pointer here. `file_mutexes` will always have a pointer, so the object will never be deleted.

Reply:
I went with the `shared_ptr` so the destructor could use the reference count to know when it is safe to remove the mutex from `file_mutexes`.  I wanted to try to keep `file_mutexes` from growing too large.  If thread 1 and 2 are both about to load file X into memory, thread 1 creates the mutex and then 2 would destroy it (assuming no other thread comes in before 2 finishes its work).  Is this making things too complicated?

Reply:
Ah, OK. I didn't look closely enough at the destructor.

I am a bit worried about this, though. It does seem pretty complicated, though it may be necessary complication.

Reply:
Yeah I'm not in love with this idea and I'm open to suggestions.  This seemed much less invasive than other ideas I could come up with.  One other idea was altering the auto-generated code for load to lock and then re-check a cache (if provided as an argument) and immediately add the loaded file to the cache before releasing the lock.

Reply:
Some (possibly bad) ideas I came up with:
- Just have a single mutex for all files. We'd only be able to read one file at a time, but that's probably what we do anyway, and reading two files from the same disk at the same time can perform badly.
- Can we do something more optimistic? Maybe we read the file unconditionally but we don't put it in the cache if it's already there.

Reply:
I like either of those better than my current implementation since it makes everything much less complicated.  I'm leaning toward using your first idea of a single load lock.

Reply:
My only concern with the single lock is that reading only a single file at once might be a performance issue, especially on high performance storage systems. However, it seems like the disk IO should be a bigger bottleneck than the serialization.